libcudacxx/include/cuda/std/__bit/byteswap.h

```diff
@@ -21,147 +21,160 @@
-  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Full __impl_recursive(_Full __val) noexcept
-  {
-    static_assert(sizeof(_Full) == sizeof(_Half) * 2, "Invalid half type passed to __bytswap_impl");
+  using _Half            = __make_nbit_uint_t<numeric_limits<_Full>::digits / 2>;
+  constexpr auto __shift = numeric_limits<_Half>::digits;
 
-    return static_cast<_Full>(__impl(static_cast<_Half>(__val >> CHAR_BIT * sizeof(_Half))))
-         | (static_cast<_Full>(__impl(static_cast<_Half>(__val))) << CHAR_BIT * sizeof(_Half));
+  if constexpr (sizeof(_Full) > 2)
+  {
+    return static_cast<_Full>(_CUDA_VSTD::__byteswap_impl(static_cast<_Half>(__val >> __shift)))
+         | (static_cast<_Full>(_CUDA_VSTD::__byteswap_impl(static_cast<_Half>(__val))) << __shift);
   }
+  else
+  {
+    return static_cast<_Full>((__val << __shift) | (__val >> __shift));
+  }
+}
 
-#if _CCCL_HAS_CUDA_COMPILER
-  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI _CCCL_DEVICE static uint32_t __impl_device(uint32_t __val) noexcept
+template <class _Tp>
+_CCCL_NODISCARD _CCCL_HIDE_FROM_ABI _CCCL_DEVICE _Tp __byteswap_impl_device(_Tp __val) noexcept
+{
+  if constexpr (sizeof(_Tp) == sizeof(uint16_t))
   {
-    uint32_t __result;
-    asm("prmt.b32 %0, %1, 0, 0x0123;" : "=r"(__result) : "r"(__val));
-    return __result;
+    return static_cast<uint16_t>(_CUDA_VPTX::prmt(static_cast<uint32_t>(__val), uint32_t{0}, uint32_t{0x3201}));
   }
+  else if constexpr (sizeof(_Tp) == sizeof(uint32_t))
+  {
+    return _CUDA_VPTX::prmt(__val, uint32_t{0}, uint32_t{0x0123});
+  }
+  else if constexpr (sizeof(_Tp) == sizeof(uint64_t))
+  {
+    const auto __hi     = static_cast<uint32_t>(__val >> 32);
+    const auto __lo     = static_cast<uint32_t>(__val);
+    const auto __new_lo = _CUDA_VPTX::prmt(__hi, uint32_t{0}, uint32_t{0x0123});
+    const auto __new_hi = _CUDA_VPTX::prmt(__lo, uint32_t{0}, uint32_t{0x0123});
 
-  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI _CCCL_DEVICE static uint64_t __impl_device(uint64_t __val) noexcept
+    return static_cast<uint64_t>(__new_hi) << 32 | static_cast<uint64_t>(__new_lo);
+  }
+  else
   {
-    uint32_t __hi;
-    uint32_t __lo;
-    asm("mov.b64 {%0, %1}, %2;" : "=r"(__hi), "=r"(__lo) : "l"(__val));
-    asm("prmt.b32 %0, %0, 0, 0x0123;" : "+r"(__hi));
-    asm("prmt.b32 %0, %0, 0, 0x0123;" : "+r"(__lo));
+    return _CUDA_VSTD::__byteswap_impl_recursive(__val);
+  }
+}
 
-    uint64_t __result;
-    asm("mov.b64 %0, {%1, %2};" : "=l"(__result) : "r"(__lo), "r"(__hi));
+template <class _Tp>
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr _Tp __byteswap_impl(_Tp __val) noexcept
+{
+  constexpr auto __shift = numeric_limits<uint8_t>::digits;
 
-    return __result;
-  }
-#endif // _CCCL_HAS_CUDA_COMPILER
+  _Tp __result{};
 
-public:
-  template <class _Tp>
-  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Tp __impl(_Tp __val) noexcept
+  _CCCL_PRAGMA_UNROLL_FULL()
+  for (size_t __i{}; __i < sizeof(_Tp); ++__i)
```

Comment:
```suggestion
  for (size_t __i = 0; __i < sizeof(_Tp); ++__i)
```
for consistency and readability